File: src/test/java/application/browser/ActionHelper.java
package application.browser;

import com.matag.player.PlayerType;
import org.openqa.selenium.By;
import org.openqa.selenium.support.ui.ExpectedConditions;

import static com.matag.player.PlayerType.OPPONENT;
import static com.matag.player.PlayerType.PLAYER;
import static org.openqa.selenium.support.ui.ExpectedConditions.elementToBeClickable;

public class ActionHelper {
  private MatagBrowser matagBrowser;

  ActionHelper(MatagBrowser matagBrowser) {
    this.matagBrowser = matagBrowser;
  }

  public void canContinue() {
    matagBrowser.wait(elementToBeClickable(getContinueSelector()));
  }

  public void cannotContinue() {
    matagBrowser.wait(ExpectedConditions.not(elementToBeClickable(getContinueSelector())));
  }

  public void clickContinue() {
    matagBrowser.getPhaseHelper().isPriority(PLAYER);
    canContinue();
    matagBrowser.findElement(getContinueSelector()).click();
  }

  public void clickContinueAndExpectPhase(String expectStatus, PlayerType player) {
    clickContinue();
    matagBrowser.player1().getPhaseHelper().is(expectStatus, player);
    matagBrowser.player2().getPhaseHelper().is(expectStatus, other(player));
  }

  private By getContinueSelector() {
    return By.id("continue-button");
  }

  private PlayerType other(PlayerType player) {
    if (PLAYER == player) {
      return OPPONENT;
    } else {
      return PLAYER;
    }
  }
}

File: src/test/kotlin/application/browser/ActionHelper.kt
package application.browser

import com.matag.player.PlayerType
import org.openqa.selenium.By
import org.openqa.selenium.support.ui.ExpectedConditions

class ActionHelper internal constructor(private val matagBrowser: MatagBrowser) {
    fun canContinue() {
        matagBrowser.wait(ExpectedConditions.elementToBeClickable(this.continueSelector))
    }

    fun cannotContinue() {
        matagBrowser.wait(ExpectedConditions.not(ExpectedConditions.elementToBeClickable(this.continueSelector)))
    }

    fun clickContinue() {
        matagBrowser.getPhaseHelper().isPriority(PlayerType.PLAYER)
        canContinue()
        matagBrowser.findElement(this.continueSelector).click()
    }

    fun clickContinueAndExpectPhase(expectStatus: String?, player: PlayerType?) {
        clickContinue()
        matagBrowser.player1().getPhaseHelper().`is`(expectStatus, player)
        matagBrowser.player2().getPhaseHelper().`is`(expectStatus, other(player))
    }

    private val continueSelector: By
        get() = By.id("continue-button")

    private fun other(player: PlayerType?): PlayerType {
        if (PlayerType.PLAYER == player) {
            return PlayerType.OPPONENT
        } else {
            return PlayerType.PLAYER
        }
    }
}
